core/txpool/blobpool/limbo.go

```diff
@@ -0,0 +1,253 @@
+		log.Error("Dropping duplicate blob limbo entry", "owner", item.TxHash, "id", id)
+		return errors.New("duplicate blob")
+	}
+	l.index[item.TxHash] = id
+
+	if _, ok := l.groups[item.Block]; !ok {
+		l.groups[item.Block] = make(map[uint64]common.Hash)
+	}
+	l.groups[item.Block][id] = item.TxHash
+
+	return nil
+}
+
+// finalize evicts all blobs belonging to a recently finalized block or older.
+func (l *limbo) finalize(final *types.Header) {
+	// Just in case there's no final block yet (network not yet merged, weird
+	// restart, sethead, etc), fail gracefully.
+	if final == nil {
+		log.Error("Nil finalized block cannot evict old blobs")
+		return
+	}
+	for block, ids := range l.groups {
+		if block > final.Number.Uint64() {
+			continue
+		}
+		for id, owner := range ids {
+			if err := l.store.Delete(id); err != nil {
+				log.Error("Failed to drop finalized blob", "block", block, "id", id, "err", err)
+			}
+			delete(l.index, owner)
+		}
+		delete(l.groups, block)
+	}
+}
+
+// push stores a new blob transaction into the limbo, waiting until finality for
+// it to be automatically evicted.
+func (l *limbo) push(tx *types.Transaction, block uint64) error {
+	// If the blobs are already tracked by the limbo, consider it a programming
+	// error. There's not much to do against it, but be loud.
+	if _, ok := l.index[tx.Hash()]; ok {
+		log.Error("Limbo cannot push already tracked blobs", "tx", tx)
+		return errors.New("already tracked blob transaction")
+	}
+	if err := l.setAndIndex(tx, block); err != nil {
+		log.Error("Failed to set and index limboed blobs", "tx", tx, "err", err)
+		return err
+	}
+	return nil
+}
+
+// pull retrieves a previously pushed set of blobs back from the limbo, removing
+// it at the same time. This method should be used when a previously included blob
+// transaction gets reorged out.
+func (l *limbo) pull(tx common.Hash) (*types.Transaction, error) {
+	// If the blobs are not tracked by the limbo, there's not much to do. This
+	// can happen for example if a blob transaction is mined without pushing it
+	// into the network first.
+	id, ok := l.index[tx]
+	if !ok {
+		log.Trace("Limbo cannot pull non-tracked blobs", "tx", tx)
+		return nil, errors.New("unseen blob transaction")
+	}
+	item, err := l.getAndDrop(id)
+	if err != nil {
+		log.Error("Failed to get and drop limboed blobs", "tx", tx, "id", id, "err", err)
+		return nil, err
+	}
+	return item.Tx, nil
+}
+
+// update changes the block number under which a blob transaction is tracked. This
+// method should be used when a reorg changes a transaction's inclusion block.
+//
+// The method may log errors for various unexpected scenarios but will not return
+// any of it since there's no clear error case. Some errors may be due to coding
+// issues, others caused by signers mining MEV stuff or swapping transactions. In
+// all cases, the pool needs to continue operating.
+func (l *limbo) update(txhash common.Hash, block uint64) {
```

Comment:
This function can be optimized, we want update the block information in limboBlob, so we may just update the data when the rlp encoded sized does not change without deleting the old one and inserting the new old. However, the complexity might out-weight the performance gain. Just note this here in case we find reorg case big performance overhead.